gcamwrapper/query_library.py

```diff
@@ -0,0 +1,251 @@
+        plus_op = '+' if is_get_data else ''
+    ret = '[' + plus_op + 'NamedFilter,'
+    if param_operands is None or len(param_operands) == 0 or param_operands[0] == '*':
+        ret += wrapper_to_fusion_lookup['*']
+    elif not is_get_data and plus_op == '+':
+        if len(param_operands) < 1:
+            raise Exception('Invalid query parameter spec: ', param_operands)
+        ret += wrapper_to_fusion_lookup[param_operands[0]]
+    elif len(param_operands) == 2 and param_operands[0] in wrapper_to_fusion_lookup.keys():
+        ret += wrapper_to_fusion_lookup[param_operands[0]] + ',' + param_operands[1]
+    else:
+        raise Exception('Invalid query parameter spec: ', param_operands)
+    ret += ']'
+    return ret
+
+def apply_query_params(query, query_params, is_get_data):
+    '''Translate a raw query looked up from a query library into a GCAM Fusion query.
+       Raw queries will contain placeholders such as \code{'{arg_tag@arg_type}'} and those
+       will get replaced user supplied filter options supplied as a list in \code{query_params
+       where we match those keys to `arg_type`.  If `arg_type` is "name" then `parse_str_query_param`
+       will be used to process the value of query_params[[arg_tag]] and if "year" then
+       `parse_int_query_param` is used.
+       Note symantics are slightly different if is_get_data is true as described in
+       parse_.*_query_params functions.
+       For any arg_tag which has no entry in query_params it will be replaced with nothing
+       which tells GCAM Fusion to match any but do not read/write to the DataFrame for that container.
+
+    :param query: The raw query which needs to have it's placeholders translated.
+    :type query: str
+    :param query_params: The user options provided as a list of arg_tags to and array of
+                         operators and potentially operands which will get translated to GCAM
+                         Fusion syntax.
+    :type query_params: dict of str to array
+    :param is_get_data: A boolean if true follows get data symantics and set data if false.
+    :type is_get_data: boolean
+
+    :returns: A translated query into GCAM Fusion syntax that is ready to run.
+    '''
+
+    placeholders = find_placeholders(query)
+    parsed_params = dict()
+    for param, args in query_params.items():
+        if not param in placeholders.keys():
+            warnings.warn(param+' has no placeholders in '+query)
+        else:
+            # note error checking on placeholder types has already occurred
+            parsed_params[param+'@'+placeholders[param]] = parse_int_query_param(args, is_get_data) if placeholders[param] == 'year' else parse_str_query_param(args, is_get_data)
+
+    # TODO: better syntax?
+    for param, ptype in placeholders.items():
+        if not param in query_params.keys():
+            parsed_params[param+'@'+ptype] = ''
+
+    return query.format(**parsed_params)
+
+
+if __name__ == '__main__':
```

Comment:
Need to dump this from this script in the package.  We will put this in the docs